### PR TITLE
Use "discrete-time variable" consistently.

### LINKED ISSUE #2640: What kind of 'discrete' variables are initialized to pre-values in algorithms?
In [11.1.2 _Execution of an algorithm in a model_](https://specification.modelica.org/master/statements-and-algorithm-sections.html#execution-of-an-algorithm-in-a-model) we have:
> A discrete variable `v` is initialized with `pre(v)`.

What kind of _discrete_ (variable) is meant here?
- A discrete-time variable.
- A variable assigned in a when-clause.  (Subset of all discrete-time variables.)
- A variable declared `discrete`.  (Subset of all variables assigned in a when-clause.  This alternative appears the most unlikely to me.)

For illustration, consider this model:
```
model StartOrPreAlgorithmInit
  Integer x(start = 1, fixed = false) "Discrete-time variable not assigned in when-clause";
algorithm
  if 1 < time and time < 2 then
    x := 2;
  end if;
  /* What will x be after time 2? */
end StartOrPreAlgorithmInit;
```

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #2640

### CODE REVIEW

File: chapters/statements.tex [hunk 1]
```diff
@@ -33,10 +33,10 @@ \subsection{Execution of an algorithm in a model}\label{execution-of-an-algorith
 \lstinline!:=! are initialized (at least conceptually):
 \begin{itemize}
 \item
-  A non-discrete variable is initialized with its start value (i.e.\ the
+  A continuous-time variable is initialized with its start value (i.e.\ the
```
Comment (henrikt-ma) [inline chapters/statements.tex:36]: Aren't we missing parameters in initial algorithms now?

File: chapters/equations.tex [hunk 2]
```diff
@@ -555,7 +555,7 @@ \section{Synchronous Data-flow Principle and Single Assignment Rule}\label{synch
 Modelica is based on the synchronous data flow principle and the single
 assignment rule, which are defined in the following way:
 \begin{enumerate}
-\item Discrete variables keep their actual values until these variables are explicitly changed.
+\item Discrete-time variables keep their actual values until these variables are explicitly changed.
```
Comment (henrikt-ma) [inline chapters/equations.tex:558]: Shouldn't this be _Variables assigned in when-clauses_?

Comment (henrikt-ma) [inline chapters/equations.tex:558]: By the way, the use of _actual_ here seems redundant — what are we trying to say, if anything?

**Review by henrikt-ma — APPROVED** [review]
Looks good overall to me, but I think it is sufficiently non-trivial to deserve a poll.

File: chapters/statements.tex (hunk 1 above)
Comment (HansOlsson) [reply chapters/statements.tex:36]: I agree that it remains an issue, but I wouldn't put it at that line.

File: chapters/equations.tex (hunk 2 above)
Comment (HansOlsson) [reply chapters/equations.tex:558]: No, the previous statement also applied to other Integer/Boolean variables and I didn't want to change that.
```
 Integer i;
algorithm 
 if sin(time)>0.9 then
   i:=1;
 end if;
 if sin(time)<-0.9 then
   i:=-1;
 end if;
```
As far as I recall there are models relying on this. One could make some argument that this is slightly different; (the algorithm implicitly sets the variable to its pre) but that seems contrived.

Comment (henrikt-ma) [reply chapters/equations.tex:558]: Sure.  Resolving.

File: chapters/statements.tex (hunk 1 above)
Comment (henrikt-ma) [reply chapters/statements.tex:36]: So maybe add another item then?

File: chapters/equations.tex (hunk 2 above)
Comment (henrikt-ma) [reply chapters/equations.tex:558]: OK, then I suggest that we simply remove _actual_.

Comment (HansOlsson) [reply chapters/equations.tex]: Ok, done.

File: chapters/statements.tex (hunk 1 above)
Comment (HansOlsson) [reply chapters/statements.tex:36]: Done. But it required a bit more text.

File: chapters/statements.tex [hunk 3]
```diff
@@ -41,6 +41,12 @@ \subsection{Execution of an algorithm in a model}\label{execution-of-an-algorith
   If at least one element of an array appears on the left hand side of
   the assignment operator, then the complete array is initialized in
   this algorithm section.
+\item
+  A parameter assigned in an initial algorithm, \cref{initialization-initial-equation-and-initial-algorithm},
+  is initialized with its start-value (i.e.\ the value of the start-attribute)
```
Comment (henrikt-ma) [inline chapters/statements.tex:46]: I think we want `\lstinline` at least on the attribute:
```suggestion
  is initialized with its start-value (i.e.\ the value of the \lstinline!start! attribute)
```

File: chapters/statements.tex [hunk 4]
```diff
@@ -41,6 +41,12 @@ \subsection{Execution of an algorithm in a model}\label{execution-of-an-algorith
...
+  provided the start-value can be evaluated before evaluating the initial algorithm
+  (a literal value can always be evaluated); otherwise a default start-value is used
+  as if the start-attribute was not set.
```
Comment (henrikt-ma) [inline chapters/statements.tex:49]: Similarly here:
```suggestion
  as if the \lstinline!start! attribute was not set.
```

File: chapters/statements.tex [hunk 5]
```diff
@@ -41,6 +41,12 @@ \subsection{Execution of an algorithm in a model}\label{execution-of-an-algorith
   If at least one element of an array appears on the left hand side of
   the assignment operator, then the complete array is initialized in
   this algorithm section.
+\item
+  A parameter assigned in an initial algorithm, \cref{initialization-initial-equation-and-initial-algorithm},
+  is initialized with its start-value (i.e.\ the value of the start-attribute)
+  provided the start-value can be evaluated before evaluating the initial algorithm
+  (a literal value can always be evaluated); otherwise a default start-value is used
```
Comment (henrikt-ma) [inline chapters/statements.tex:48]: This doesn't look as simple as I expected.  Why is it necessary to have the initialization conditioned on whether the start-value can be initialized before the initial algorithm?  To me, that just sounds like badly entangled initialization.

Comment (HansOlsson) [reply chapters/statements.tex:48]: I agree that it is a badly entangled initialization (i.e. a model issue), and we should state that more clearly. (I had in mind that an iteration variable without start-value normally leads to diagnostics, but then I realized that people cannot read my mind.)

However, it should be rare and even when it used it is not always the case that the start-values are that critical, and there are two possibilities for handling them - either we forbid it or we add a fall-back as I did above.

The problem with forbidding it is that we currently state that we conceptually always initialize variables at the beginning of each algorithm, and leave it to the tools to skip when not needed. If we want to forbid the entangled mess that might be a bit excessive - and we would then need to more clearly define when.

File: chapters/statements.tex [hunk 6]
```diff
@@ -46,7 +46,7 @@ \subsection{Execution of an algorithm in a model}\label{execution-of-an-algorith
   is initialized with its start-value (i.e.\ the value of the \lstinline!start! attribute)
   provided the start-value can be evaluated before evaluating the initial algorithm
   (a literal value can always be evaluated); otherwise a default start-value is used
-  as if the \lstinline!start! attribute was not set.
+  (together with appropriate diagnostics) as if the \lstinline!start! attribute was not set.
```
Comment (henrikt-ma) [inline chapters/statements.tex:49]: I don't find this sufficient.  If one wants the _appropriate diagnostics_, I think one should rely on a non-stringent tool that to deliver non-standard behavior (together with appropriate diagnostics).  From the specification's point of view, on the other hand, I'd say treating the inability to apply provided `start` attributes as an error is the only reasonable alternative.

File: chapters/statements.tex (hunk 5 above)
Comment (henrikt-ma) [reply chapters/statements.tex:48]: I see it differently.  What you propose is in just as much need of a clearly defined _when_ as the error approach, but the error approach has the advantage that any tool that is able to respect applicable `start` attributes should get the same simulation result.

Assuming that we really want to be able to accept these corner case models at all, it becomes a quality of implementation to do the symbolic processing that proves that a nasty `start` doesn't even come into play, but that's a smaller problem than having simulation results where the values coming out of the initialization is a quality of implementation.

Done this way, when one encounters a model developed by another tool where the `start` cannot be evaluated, this should be interpreted as that the other tool was either able to do the evaluation, or that the other tool was able to prove that the `start` doesn't come into play.  A less careful tool would then guess the latter, give a warning about it, and proceed, but this shouldn't be described in (the normative parts of) the specification.

**Comment by HansOlsson** [comment]
Agreement at phone meeting.

File: chapters/statements.tex (hunk 5 above)
Comment (HansOlsson) [reply chapters/statements.tex:48]: But in the relevant cases the ignored start-value is a guess-value, and there are multiple solutions to the initialization problem (on top of the other mess), and we cannot guarantee that tools produce the same solution even for the same guess-value.

Comment (henrikt-ma) [reply chapters/statements.tex:48]: I think this part should have been excluded when resolving #2640, as we didn't discuss this part at the phone meeting.  I'll make a PR for removing this part, and open an issue to continue the discussion.

**Pull request merged**